src/main/java/ui/TestController.java

```diff
@@ -79,4 +85,46 @@ public static boolean isCloseOnQuit() {
         return hasUI() && commandLineArgs.getOrDefault("closeonquit", "false").equalsIgnoreCase("true");
     }
 
+    /**
+     * Creates a Preferences that store data in test config file if run with test mode
+     * or in a default config file specified in the Preferences class
+     * @return
+     */
+    public static Preferences createApplicationPreference() {
+        if (isTestMode()) {
+            return createTestPreferences();
+        }
+
+        return new Preferences(Preferences.GLOBAL_CONFIG_FILE);
+    }
+
+    /**
+     * Creates a Preferences that stores data in the test directory
+     * @return
+     */
+    public static Preferences createTestPreferences() {
+        return new Preferences(Preferences.TEST_CONFIG_FILE);
+    }
+
+    /**
+     * Creates a RepoIO for the application that uses different components
+     * depending on various test options: --test, --testjson etc.
+     * @return
+     */
+    public static RepoIO createApplicationRepoIO() {
+        return new RepoIO(isTestMode() ? Optional.of(new DummySource()) : Optional.empty(),
+                          isTestJSONEnabled() ? Optional.of(new JSONStoreStub()) : Optional.empty(),
```

Comment:
I am confused about this line. In the original, if `isTestJSONEnabled()` is true, it will use `Optional.empty`. Otherwise, it will use the stub. This holds true if it is in test mode.

I think this can be made close to the original and more readable with this
```
if (isTestMode()) {
  return new RepoIO(Optional.of(new DummySource()),
      isTestJSONEnabled() ? Optional.empty() : Optional.of(new JSONStoreStub()),
      Optional.of(RepoStore.TEST_DIRECTORY));
} else {
  return new RepoIO(Optional.empty(), Optional.empty(), Optional.empty());
}
```

The above should be safe since if test json command line option is set to true, i.e. `isTestJSONEnabled()` is true, `isTestMode()` would return true. Hence if developer forgot to put test mode to true when testing json is set to true, the real JSONs will not be messed up.